### PR TITLE
bugfix: allow lazy load within transaction

### PR DESCRIPTION
**Pull request by DR-BoneZ**

### CODE REVIEW

**Comment by DR-BoneZ** [comment]
broken by #1954

File: src/find-options/operator/Raw.ts [hunk 1]
```diff
@@ -4,6 +4,6 @@ import {FindOperator} from "../FindOperator";
  * Find Options Operator.
  * Example: { someField: Raw([...]) }
  */
-export function Raw<T>(value: string|((columnAlias?: string) => string)) {
-    return new FindOperator("raw", value as any, false);
+export function Raw<T>(value: string|((columnAlias?: string) => string)): any { // TODO: make this typecheck
```
Comment (pleerock) [inline src/find-options/operator/Raw.ts]: why did you change it? There are no compiler errors we currently have.

File: src/metadata/EntityMetadata.ts [hunk 2]
```diff
@@ -485,14 +486,14 @@ export class EntityMetadata {
     /**
      * Creates a new entity.
      */
-    create(): any {
+    create(queryRunner: QueryRunner): any {
```
Comment (pleerock) [inline src/metadata/EntityMetadata.ts]: `queryRunner` should be optional. Don't pass it using `createQueryRunner()` into the method in the places **where you don't need to do it**.

File: test/functional/relations/lazy-relations/basic-lazy-relation/basic-lazy-relations.ts [hunk 3]
```diff
@@ -326,4 +326,23 @@ describe("basic-lazy-relations", () => {
         loadedPost.title.should.be.equal("post with great category");
     })));
 
+    it("should successfully load relations within a transaction", () => Promise.all(connections.map(async connection => {
+        if (connection.driver.options.type !== "postgres") return;
```
Comment (pleerock) [inline test/functional/relations/lazy-relations/basic-lazy-relation/basic-lazy-relations.ts]: why did you skip other drivers?

Comment (DR-BoneZ) [reply test/functional/relations/lazy-relations/basic-lazy-relation/basic-lazy-relations.ts]: this is temporary for testing locally

File: src/find-options/operator/Raw.ts (hunk 1 above)
Comment (DR-BoneZ) [reply src/find-options/operator/Raw.ts]: strange.. I'm running into a compiler warning saying T is unused. I can revert this if this error is ignored.

Comment (pleerock) [reply src/find-options/operator/Raw.ts]: make sure you use correct typescript version. There are no typescript errors in CI and my local machine

File: src/metadata/EntityMetadata.ts [hunk 4]
```diff
@@ -485,14 +486,18 @@ export class EntityMetadata {
     /**
      * Creates a new entity.
      */
-    create(): any {
+    create(queryRunner: QueryRunner = this.connection.createQueryRunner()): any {
```
Comment (pleerock) [inline src/metadata/EntityMetadata.ts]: no! it should be optional means you should be `queryRunner?: QueryRunner`. Creating a query runner is a responsible task, don't do that until you absolutely need it.

File: src/metadata-builder/EntityMetadataBuilder.ts [hunk 5]
```diff
@@ -209,7 +209,7 @@ export class EntityMetadataBuilder {
                 entityMetadata.relations
                     .filter(relation => relation.isLazy)
                     .forEach(relation => {
-                        this.connection.relationLoader.enableLazyLoad(relation, (entityMetadata.target as Function).prototype);
+                        this.connection.relationLoader.enableLazyLoad(relation, (entityMetadata.target as Function).prototype, this.connection.createQueryRunner());
```
Comment (pleerock) [inline src/metadata-builder/EntityMetadataBuilder.ts]: queryRunner should be optional inside `enableLazyLoad` and you should not create query runner here.

File: src/query-builder/SelectQueryBuilder.ts [hunk 6]
```diff
@@ -1819,10 +1819,11 @@ export class SelectQueryBuilder<Entity> extends QueryBuilder<Entity> implements
 
         if (rawResults.length > 0) {
 
+            const queryRunner = this.obtainQueryRunner();
             // transform raw results into entities
             const rawRelationIdResults = await relationIdLoader.load(rawResults);
             const rawRelationCountResults = await relationCountLoader.load(rawResults);
-            const transformer = new RawSqlResultsToEntityTransformer(this.expressionMap, this.connection.driver, rawRelationIdResults, rawRelationCountResults);
+            const transformer = new RawSqlResultsToEntityTransformer(this.expressionMap, this.connection.driver, rawRelationIdResults, rawRelationCountResults, queryRunner);
```
Comment (pleerock) [inline src/query-builder/SelectQueryBuilder.ts]: take a look how in all other places where `this.obtainQueryRunner()` is called resulted query runner is released if it was created by query runner. And it should be done only once because once query runner is released you can't use it anymore.

I think proper solution here is to make queryRunner inside `RawSqlResultsToEntityTransformer` constructor optional and pass `this. queryRunner` here.

Sorry for being annoying about query runners here, but they are very dangerous if are not used properly and can lead to leaks and hard-to-debug bugs.

Comment (DR-BoneZ) [reply src/query-builder/SelectQueryBuilder.ts]: yeah no worries. still getting the hang of the setup.

**Comment by DR-BoneZ** [comment]
after all hook is timing out. not sure if this is related to my changes?

File: src/entity-manager/EntityManager.ts [hunk 7]
```diff
@@ -473,7 +473,7 @@ export class EntityManager {
      */
     async find<Entity>(entityClass: ObjectType<Entity>|EntitySchema<Entity>|string, optionsOrConditions?: FindManyOptions<Entity>|FindConditions<Entity>): Promise<Entity[]> {
         const metadata = this.connection.getMetadata(entityClass);
-        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name);
+        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name, this.queryRunner);
```
Comment (pleerock) [inline src/entity-manager/EntityManager.ts]: I think you don't need to pass `this.queryRunner` here since `this.createQueryBuilder` already uses it.

File: src/entity-manager/EntityManager.ts [hunk 8]
```diff
@@ -502,7 +502,7 @@ export class EntityManager {
      */
     async findAndCount<Entity>(entityClass: ObjectType<Entity>|EntitySchema<Entity>|string, optionsOrConditions?: FindManyOptions<Entity>|FindConditions<Entity>): Promise<[Entity[], number]> {
         const metadata = this.connection.getMetadata(entityClass);
-        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name);
+        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name, this.queryRunner);
```
Comment (pleerock) [inline src/entity-manager/EntityManager.ts]: I think you don't need to pass `this.queryRunner` here since `this.createQueryBuilder` already uses it.

File: src/entity-manager/EntityManager.ts [hunk 9]
```diff
@@ -532,7 +532,7 @@ export class EntityManager {
         if (!ids.length)
             return Promise.resolve([]);
         const metadata = this.connection.getMetadata(entityClass);
-        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name);
+        const qb = this.createQueryBuilder(entityClass, FindOptionsUtils.extractFindManyOptionsAlias(optionsOrConditions) || metadata.name, this.queryRunner);
```
Comment (pleerock) [inline src/entity-manager/EntityManager.ts]: I think you don't need to pass `this.queryRunner` here since `this.createQueryBuilder` already uses it.

File: src/entity-manager/EntityManager.ts [hunk 10]
```diff
@@ -580,7 +580,7 @@ export class EntityManager {
         } else if (maybeOptions && FindOptionsUtils.isFindOneOptions(maybeOptions) && maybeOptions.join) {
             alias = maybeOptions.join.alias;
         }
-        const qb = this.createQueryBuilder(entityClass, alias);
+        const qb = this.createQueryBuilder(entityClass, alias, this.queryRunner);
```
Comment (pleerock) [inline src/entity-manager/EntityManager.ts]: I think you don't need to pass `this.queryRunner` here since `this.createQueryBuilder` already uses it.

File: src/query-builder/SelectQueryBuilder.ts (hunk 6 above)
Comment (pleerock) [reply src/query-builder/SelectQueryBuilder.ts]: You still have an issue here (use this.queryRunner and remove `const queryRunner = this.obtainQueryRunner();`). Maybe you'll success with tests if you fix this one.

Comment (DR-BoneZ) [reply src/query-builder/SelectQueryBuilder.ts]: That was it. Thanks!

**Comment by pleerock** [comment]
Looks like everything is fine and we can merge this PR

**Comment by pleerock** [comment]
Thank you very much!
